### PR TITLE
[ONNX][TypePromo] Explicit type promotion pass

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104491
* #104720
* #104229
* __->__ #104064

This PR adds the `ExplicitTypePromotionPass` that does an fx graph to fx graph transformation
explicitly adding cast nodes into the graph to emulate the PyTorch type promotion behavior.

Full design doc and discussion at https://microsoft-my.sharepoint.com/:w:/p/bowbao/Edj2lF1oi0JIitT_3ntyuqkBo6ll7N6NJDmavM0lp_KkEA?e=OElyjR

### CODE REVIEW

File: test/onnx/test_fx_op_consistency.py [hunk 1]
```diff
@@ -556,7 +490,7 @@
         and sample.kwargs.get("padding") == 1,
         reason="FIXME: After https://github.com/microsoft/onnxruntime/issues/15446 is fixed",
     ),
-    xfail(
+    skip(
```
Comment (BowenBao) [inline test/onnx/test_fx_op_consistency.py:493]: After type promotion this test starts to fail a few subtests with unexpected pass, which appears to be incosistent with the taggex xfail reason.

File: .ci/docker/common/install_onnx.sh [hunk 2]
```diff
@@ -24,7 +24,7 @@ pip_install \
   transformers==4.25.1
 
 # TODO: change this when onnx-script is on testPypi
-pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@7e131c578f290ffad1f26bacda11a83daf5476ba"
+pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738"
```
Comment (BowenBao) [inline .ci/docker/common/install_onnx.sh:27]: note: for missing `aten.tensor` op support that is used when promoting python number to tensor.

File: torch/onnx/_internal/diagnostics/_rules.py [hunk 3]
```diff
@@ -844,6 +864,31 @@ class _POERules(infra.RuleCollection):
     )
     """Result from FX graph analysis to reveal unsupported FX nodes."""
 
+    fx_node_explicit_type_promotion: _FxNodeExplicitTypePromotion = dataclasses.field(
+        default=_FxNodeExplicitTypePromotion.from_sarif(
+            **{
+                "id": "FXE0013",
```
Comment (titaiwangms) [inline torch/onnx/_internal/diagnostics/_rules.py:870]: CONFLICT!!! FXE0013 has been taken. By the way, what does FXE mean? If it only means FX exporter, do we want to utilize it to further separate rules under FX exporter?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 4]
```diff
@@ -20,15 +21,26 @@
...
 
 from torch.onnx._internal import _beartype
-from torch.utils import _pytree
+from torch.onnx._internal.fx import _pass, diagnostics
+from torch.utils import _python_dispatch, _pytree
 
 logger = logging.getLogger(__name__)
 
+# TODO(bowbao): move to type utils.
+_SCALAR_TYPE_TENSOR_DTYPE_MAP: Mapping[type, torch.dtype] = {
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/passes/type_promotion.py:37]: Also please make it accessible through a stateless function or class.

Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/type_promotion.py:37]: Maybe we can unify the need of torch.dtype conversion into something like JitScalarType system now. The needs are increasing.

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 5]
```diff
@@ -313,6 +313,32 @@
...
+      process in PyTorch that it automatically promotes the input types of an operator to a common type.
+      The common type is determined by the type promotion rule, and each operator has its own type promotion rules.
+      To learn more about PyTorch type promotion rules, checkout [elementwise_dtypes doc](https://github.com/pytorch/pytorch/blob/f044613f78df713fb57f70c608483c9f10ad332e/torch/_prims_common/__init__.py#L1252-L1335),
+      and [torch._refs ops](https://github.com/pytorch/pytorch/blob/a475ea4542dfe961c9d097e33ab5041f61c8c17f/torch/_refs/__init__.py#L484).
+
+      In ONNX, implicit type promotion is not supported. Therefore, we need to explicitly insert cast nodes to emulate
+      the PyTorch behavior. This diagnostic tracks the process of node level explicit type promotion.
+
+      A set of type promotion rules supported by this pass is materialized under `torch/onnx/_internal/fx/passes/type_promotion.py`.
+      Checkout [Note: Add type promotion rule] how to update or add new type promotion rules.
```
Comment (titaiwangms) [inline torch/onnx/_internal/diagnostics/rules.yaml:333]: If it's not rephrased by chatGPT yet, you can try with it. Not necessarily, but it does give me a decent docs.

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 6]
```diff
@@ -313,6 +313,32 @@
...
+  message_strings:
+    default:
+      text: "Performing explicit type promotion for node {target}. "
```
Comment (titaiwangms) [inline torch/onnx/_internal/diagnostics/rules.yaml:336]: I forgot to ask, is {target} decided by the decorated function?

**Comment by BowenBao** [comment]
I will update new diagnostic rule id prior to merging.

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 7]
```diff
@@ -313,6 +313,32 @@
...
     tags: []
 
+- id: FXE0013
+  name: fx-node-explicit-type-promotion
+  short_description:
+    text: Determine if type promotion is required for the FX node. Explicitly insert cast nodes if needed.
+  full_description:
+    text: "Determine if type promotion is required for the FX node. Explicitly insert cast nodes if needed."
+    markdown: |
+      This diagnostic tracks the process of node level explicit type promotion. Implicit [type promotion](https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc) is a
```
Comment (titaiwangms) [inline torch/onnx/_internal/diagnostics/rules.yaml:323]: I think we should have a better explanation here on "it's needed to be explicit in ONNX, because it's implicit in PyTorch.". It seems to me the explanation here is a bit confusing.

Comment (titaiwangms) [reply torch/onnx/_internal/diagnostics/rules.yaml:323]: The same idea goes to the name itself. When we say: fx-node-explicit-type-promotion, it means we need explicit type promotion for ONNX. But it could be just me...

File: torch/onnx/_internal/exporter.py [hunk 8]
```diff
@@ -635,6 +635,10 @@ def pre_export_passes(
     # Remove them since ONNX inference does not need them.
     module = passes.RemoveInputMutation(diagnostic_context, module).run(*fx_module_args)
 
+    # ONNX does not support concept of (implicit) type promotion.
+    # Insert type casts explicitly where needed.
+    module = passes.ExplicitTypePromotionPass(diagnostic_context, module).run()
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter.py:640]: Is there a convention for naming the passes? Looks like the surrounding ones start with verbs

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/type_promotion.py:37]: > Also please make it accessible through a stateless function or class.

If it is private I think it's fine for now

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 9]
```diff
@@ -55,6 +67,16 @@ class TypePromotionSnapshot:
     """Expected output dtype of the node."""
 
 
+@_beartype.beartype
+def _get_fake_tensor_val(node: torch.fx.Node) -> fake_tensor.FakeTensor:
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:71]: Add a brief on how it should be used?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 10]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+) -> Optional[TypePromotionRule]:
+    """Get type promotion rule for a node.
+
+    Args:
+        diagnostic: Diagnostic object.
+        node: Node to get type promotion rule for.
+        type_promotion_table: Type promotion table.
+
+    Returns:
+        Type promotion rule for the node.
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:966]: Document None as well (when it happens)?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 11]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+        diagnostic.with_additional_message(
+            f"Skipped. Cannot find type promotion rule for op: {op}"
+        )
+        return None
+
+    diagnostic.with_additional_message(f"Found type promotion rule: {rule}")
+    return rule
+
+
+class _OpTraceDispatchMode(_python_dispatch.TorchDispatchMode):
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:985]: Just for my learning. What does it do/why do we need it? Thanks!

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 12]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+        super().__init__(*args, **kwargs)
+        self.traced_ops = []
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        self.traced_ops.append(func)
+        return func(*args, **kwargs)
+
+
+@_beartype.beartype
+def find_compatible_op_overload(
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:998]: So this is like a tiny dispatcher?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 13]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+) -> torch._ops.OpOverload:
+    """Find the compatible op overload for a node with a given set of args and kwargs.
+
+    Returns:
+        Compatible op overload for the node.
+
+    Raises:
+        RuntimeError: If no compatible op overload is found.
+        AssertionError: If compatible op overload is not an instance of torch._ops.OpOverload.
+        AssertionError: If op is being decomposed.
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1009]: Do we expect them to happen with valid user input? If not I don't think we need to document them here. Instead I would document as invariants under the summary, before "Returns".

Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1009]: If yes we should raise something that's not assertion error

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 14]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+        new_op_overload, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(new_op_overload)}"
+    assert (
+        new_op_overload.overloadpacket == op
+    ), f"Expected same OpOverload packet, got {new_op_overload.overloadpacket} != {op}"
+
+    return new_op_overload
+
+
+@dataclasses.dataclass
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1034]: I am here. Taking a break for now.

File: torch/onnx/_internal/diagnostics/_rules.py (hunk 3 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/diagnostics/_rules.py:870]: good point, do we have any way to find this conflict? maybe a test that runs on ci?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 15]
```diff
@@ -55,6 +67,16 @@ class TypePromotionSnapshot:
     """Expected output dtype of the node."""
 
 
+@_beartype.beartype
+def _get_fake_tensor_val(node: torch.fx.Node) -> fake_tensor.FakeTensor:
+    val = node.meta.get("val", None)
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:72]: move inside the if? it makes no sense to set it to none if it will railse on next if

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 16]
```diff
@@ -68,8 +90,11 @@ def __init__(
     ):
         self.namespace = namespace
         self.op_name = op_name
+        # Positions of args to promote.
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:93]: maybe create official docs strings with `Args:` section?

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 12 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/passes/type_promotion.py:998]: Indeed, confusing. Any chance we could reuse or extend the current our dispatcher for this?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 17]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+    ), f"Expected OpOverload, got {type(new_op_overload)}"
+    assert (
+        new_op_overload.overloadpacket == op
+    ), f"Expected same OpOverload packet, got {new_op_overload.overloadpacket} != {op}"
+
+    return new_op_overload
+
+
+@dataclasses.dataclass
+class NodeMeta:
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1035]: Are we keeping another "meta" structure or using this class to update the fx node meta field? please elaborate on the docstring

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 18]
```diff
@@ -920,3 +947,378 @@ def get_rule(
...
+    stack_trace: str
+    nn_module_stack: Any
+
+    def assign_to_node(self, node: torch.fx.Node) -> None:
+        """Assigns the meta data to the given node by updating `node.meta`."""
+        for field in dataclasses.fields(self):
+            setattr(node.meta, field.name, getattr(self, field.name))
+
+
+class ExplicitTypePromotionPass(_pass.Transform, torch.fx.Interpreter):
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1050]: wow, multiple inheritance. We certainly need a lot of docstring here to describe what and how things are being done. examples appreciated :)

**Review by thiagocrepaldi — COMMENTED** [review]
This PR has a lot of complexity and very little documentation. Hard to follow

File: test/onnx/test_fx_op_consistency.py (hunk 1 above)
Comment (BowenBao) [reply test/onnx/test_fx_op_consistency.py:493]: Actually more likely due to the bumping onnxscript version in this PR, so unrelated to type promotion. The inconsistency concern still holds.

File: torch/onnx/_internal/exporter.py (hunk 8 above)
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:640]: good catch, I will name it as `InsertExplicitTypePromotion`.

File: torch/onnx/_internal/diagnostics/_rules.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/diagnostics/_rules.py:870]: don't worry, it is merge conflicts 🤣

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:37]: Should we make function accessor `get(..., None)` style? I'm not sure but it is used here so I'm leaving out type_utils in this PR.

File: torch/onnx/_internal/diagnostics/rules.yaml (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/_internal/diagnostics/rules.yaml:323]: Is `fx-node-insert-explicit-type-promotion` better? Or just `fx-node-insert-type-promotion`

File: torch/onnx/_internal/exporter.py (hunk 8 above)
Comment (BowenBao) [reply torch/onnx/_internal/exporter.py:640]: Or would you prefer `InsertTypePromotion`?

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 9 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:71]: syntactic sugar, maybe call it `_fake_tensor_from_meta`?

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 17 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1035]: It is an attempt, but I'll remove it because this is not actually used.

It is painful to maintain this node.meta dict. The idea is to create an ONNX meta abstraction to be used with our fx.Node creation api such that when people create new nodes they don't forget to set certain meta info: every field is required.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 9 above)
Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/type_promotion.py:71]: or `_fake_tensor_from_node_meta`

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 18 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1050]: I will separate it. Thinking again, it is better at least in this PR to keep them separated.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 15 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:72]: Not sure how to do it without accessing node.meta multiple times.
With walrus I can make 1 line shorter but I don't feel it's necessary.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 12 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:998]: I'll expand the docstring. This is indeed a tricky problem. I have tried to solve in multiple different ways. 
The current one is the most elegant and simple solution.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 13 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1009]: I'll drop them from docstring.

File: torch/onnx/_internal/diagnostics/rules.yaml (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/_internal/diagnostics/rules.yaml]: I'm leaving `explicit` in most of the places and all the names.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 11 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1009]: added docstring

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 16 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py]: done

File: torch/csrc/onnx/diagnostics/generated/rules.h [hunk 19]
```diff
@@ -19,14 +19,12 @@ enum class Rule : uint32_t {
   kNodeMissingOnnxShapeInference,
 
   /**
-   * @brief Missing symbolic function for custom PyTorch operator, cannot
-   * translate node to ONNX.
+   * @brief Missing symbolic function for custom PyTorch operator, cannot translate node to ONNX.
```
Comment (BowenBao) [inline torch/csrc/onnx/diagnostics/generated/rules.h:22]: this is strange 🤔 wonder if lint config changed

Comment (BowenBao) [reply torch/csrc/onnx/diagnostics/generated/rules.h:22]: forgot to lintrunner init

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 20]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+
+    @_beartype.beartype
+    def _promote_arg(
+        self,
+        diagnostic: diagnostics.Diagnostic,
+        node: torch.fx.Node,
+        fx_arg: torch.fx.node.Argument,
+        dtype: Optional[torch.dtype],
+    ) -> torch.fx.node.Argument:
+        """Promote fx_arg to dtype if necessary."""
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1202]: Does the method check for necessity or do we call it if necessary? If the former, consider `_maybe_promote_arg`?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 21]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+            raise NotImplementedError(f"Unknown fx arg type: {type(fx_arg)}")
+
+    @_beartype.beartype
+    def _promote_node(
+        self,
+        diagnostic: diagnostics.Diagnostic,
+        node: torch.fx.Node,
+        rule: TypePromotionRule,
+    ) -> torch.fx.Node:
+        """Promote node inputs and outputs according to type promotion rule."""
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1272]: Just for my understanding, only _promote_node will use _promote_arg right?

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 22]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+        else:
+            diagnostic.message = f"Type promotion not needed for {node}. "
+
+        return node
+
+    @diagnostics.diagnose_call(
+        rule=diagnostics.rules.fx_node_insert_type_promotion,
+        level=diagnostics.levels.NONE,
+    )
+    def run_node(self, node: torch.fx.Node) -> Any:
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1306]: It'd be helpful to explain the high-level calling logic in English here I think: Is it `run_node -> _promote_node -> _promote_args -> _create_node -> _run_node_and_update_meta_val -> rerun_node`? When we do rerun, does it happen exactly once for each node?

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 12 above)
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1027]: Looks like when we type promote(change) args/kwargs, the opoverloadpacket needs to be dispatched again in FX graph.

I wonder if they have a schema matching system that we could use.

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1027]: > I wonder if they have a schema matching system that we could use.

afaik there's only a lower level api that checks if two schemas are compatible to each other, and it is painful to use in this scenario since we need to create a new schema from args/kwargs and check existing schemas for matches. And furthermore due to the ordering of check the schema we found might disagree with torch dispatcher. I found it easier to let `__torch_dispatch__` tell us which opoverload it dispatched to in the end.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 20 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1202]: I can rename both with `_maybe` prefix.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 21 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1272]: Correct.

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 22 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1306]: Yea except for swapping this order here `rerun_node -> _run_node_and_update_meta_val`. Each node is guaranteed to run at least once. We can change the code to run every node exactly once, but it adds more complexity and I don't feel it is necessary since this is always running in fake mode.

I'll update the docstring with high level logic of things to do.

Comment (justinchuby) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1306]: > Each node is guaranteed to run at least once.

Sounds good! I would just make this clear in the docs

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 23]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+    """
+
+    def __init__(
+        self,
+        diagnostic_context: diagnostics.DiagnosticContext,
+        module: torch.fx.GraphModule,
+        type_promotion_table: Optional[TypePromotionTable] = None,
+    ):
+        super().__init__(diagnostic_context, module)
+        self.type_promotion_table = type_promotion_table or TypePromotionTable()
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1344]: this can be inferred from `self.interpreter.type_promotion_table`.

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 24]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+        self.type_promotion_table = type_promotion_table
+
+    def _run_node_and_update_meta_val(self, node) -> Any:
+        """Run a node and update node.meta["val"] with the output value.
+
+        This should be used on new nodes or nodes that have been modified.
+        By default `Interpreter.run_node` does not update `node.meta["val"]`.
+        """
+        out = super().run_node(node)
+        self.env[node] = out
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1100]: ok "Update interpreter env state with new output value."

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 25]
```diff
@@ -933,3 +968,438 @@ def get_rule(
...
+        node_val = node.meta.get("val", None)
+        assert node_val is not None, f"Node {node} node.meta['val'] is not set."
+        args, kwargs = self.fetch_args_kwargs_from_env(node)
+        target = node.target
+        assert isinstance(
+            target, torch._ops.OpOverload
+        ), f"Expected OpOverload, got {type(target)}"
+        node.target = find_compatible_op_overload(target.overloadpacket, args, kwargs)
+
+        new_node_val = self._run_node_and_update_meta_val(node)
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1148]: No because this is not a new node. Frankly this is case by case. In this pass for modified nodes we only need to update 'val' in 'meta'. But if you prefer I can simplify the api to always copy all other meta info.

Largely this is the problem with freestyle design of `fx.Node` meta. Devs need to be super aware of its content to correctly maintain it. And meta is very important as it serves as source info for numerous high level apis.

[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1148]: I will isolate the change in a follow up PR.

[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1149]: https://github.com/pytorch/pytorch/pull/104720
